core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1050,7 +1050,8 @@ class Log(@volatile private var _dir: File,
                      leaderEpoch: Int,
                      origin: AppendOrigin = AppendOrigin.Client,
                      interBrokerProtocolVersion: ApiVersion = ApiVersion.latestVersion): LogAppendInfo = {
-    append(records, origin, interBrokerProtocolVersion, assignOffsets = true, leaderEpoch, ignoreRecordSize = false)
+    val assignOffsets = if (origin == AppendOrigin.Replication) false else true
```

Comment:
This can be replaced by `val assignOffsets = if (origin != AppendOrigin.Replication)`

Reply:
@dengziming Thanks, but it needs `else` for a complete `if` expression

Reply:
o, I mean `val assignOffsets = origin != AppendOrigin.Replication`

Reply:
Got it, will let @hachikuji review and fix it in one shot, thanks

Reply:
By the way, it seems we might want to rename `assignOffsets` since we also rely on this flag for record validation. It's a bit on the verbose side, but maybe `validateMessagesAndAssignOffsets` to go along with the similarly named method in `LogValidator`?

Reply:
Make sense, updated